### ISSUE TITLE
Some questions about ConcatDataset 

### ISSUE THREAD

**Issue by ZCWei51** [meta-llama/llama-recipes#430]
### 🚀 The feature, motivation and pitch

Hi~ I'm a student just starting out with codellama fine tuning, and I'm curious and have a little bit of a question about a data processing phenomenon.

In file src/llama_recipes/data/concatenator.py `Class ConcatDataset` . 
It seems to splice all the data according to the key values (`input_ids,labels,attention_mask`), for example `chunk_size=1024` and the average length of code snippets is `100`, in which case the input_ids after the splice should contain several codes (`[[code1+code2 +...]]`) . This seems inconsistent with making the code length consistent by following padding (`[[code1+padding], [code2+padding], [...]]`) , but this (`Class ConcatDataset`) works better I found after running the code translation task.

This approach is now found to take less time to train and still have better results, is there any better explanation for the effect of this approach?

**Comment by mreso** [comment]
The concatenation strategy is more efficient with respect to the GPU usage as no data is ignored (padding gets processes but is ignored). Thats why its often yield better performance. The [length based clustering strategy](https://github.com/meta-llama/llama-recipes/blob/14e4b05bedd588af22c44b4cbfb7a3f502fd9707/src/llama_recipes/configs/training.py#L15C52-L15C59) tries to minimize padding while avoiding context dilution. My understanding is that the accuracy does not suffer from concatenation approach as the additional context does not harm the fine-tuning. Pre-training is in fact done in the same way,.So the data in fine-tuning looks more similar to the data in pre-training.
Feel free to reopen if more questions arise.
